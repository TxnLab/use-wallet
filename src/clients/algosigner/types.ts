import type _algosdk from "algosdk";
import { PROVIDER_ID } from "../../constants";
import type { AlgodClientOptions, Network } from "../../types";

export type WindowExtended = { AlgoSigner: AlgoSigner } & Window &
  typeof globalThis;

export type AlgoSignerTransaction = {
  txn: string;
  // array of addresses to sign with (defaults to the sender),
  // setting this to an empty array tells AlgoSigner
  // that this transaction is not meant to be signed
  signers?: [];
  multisig?: string; // address of a multisig wallet to sign with
};

export type SupportedLedgers = "MainNet" | "TestNet" | "BetaNet" | string;

export type AlgoSigner = {
  connect: () => Promise<Record<string, never>>;
  accounts: (ledger: {
    ledger: SupportedLedgers;
  }) => Promise<{ address: string }[]>;
  signTxn: (transactions: AlgoSignerTransaction[]) => Promise<
    {
      txID: string;
      blob: string;
    }[]
  >;
  encoding: {
    msgpackToBase64(transaction: Uint8Array): string;
    byteArrayToString(transaction: Uint8Array): string;
  };
};

export type AlgoSignerClientConstructor = {
  client: AlgoSigner;
  id: PROVIDER_ID;
  algosdk: typeof _algosdk;
  algodClient: _algosdk.Algodv2;
  network: SupportedLedgers;
};

export type InitParams = {
  algodOptions?: AlgodClientOptions;
  algosdkStatic?: typeof _algosdk;
  network?: Network;
};
