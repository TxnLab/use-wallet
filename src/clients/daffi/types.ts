import type algosdk from 'algosdk'
import type { DaffiWalletConnect } from '@daffiwallet/connect'
import type { Transaction } from 'algosdk'
import type { Network } from '../../types/node'
import type { Metadata } from '../../types/wallet'

export type DaffiWalletConnectOptions = {
  bridge?: string
  shouldShowSignTxnToast?: boolean
  chainId?: 416001 | 416002 | 416003 | 4160
}

export interface DaffiTransaction {
  txn: Transaction
  /**
   * Optional list of addresses that must sign the transactions.
   * Wallet skips to sign this txn if signers is empty array.
   * If undefined, wallet tries to sign it.
   */
  signers?: string[]
}

export type DaffiWalletClientConstructor = {
  metadata: Metadata
  client: DaffiWalletConnect
  clientOptions?: DaffiWalletConnectOptions
  algosdk: typeof algosdk
  algodClient: algosdk.Algodv2
  network: Network
}
