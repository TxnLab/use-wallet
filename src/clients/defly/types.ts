import type { Transaction } from "algosdk";
import type { AlgodClientOptions, Network } from "../../types";
import type { DeflyWalletConnect } from "@blockshake/defly-connect";
import type _algosdk from "algosdk";

export type ClientOptions = {
  bridge?: string;
  deep_link?: string;
  app_meta?: {
    logo: string;
    name: string;
    main_color: string;
  };
  shouldShowSignTxnToast?: boolean;
};

export type DeflyTransaction = {
  txn: Transaction;
  /**
   * Optional list of addresses that must sign the transactions.
   * Wallet skips to sign this txn if signers is empty array.
   * If undefined, wallet tries to sign it.
   */
  signers?: string[];
};

export type DeflyWalletClientConstructor = {
  client: DeflyWalletConnect;
  algosdk: typeof _algosdk;
  algodClient: _algosdk.Algodv2;
  network: Network;
};

export type InitParams = {
  clientOptions?: ClientOptions;
  algodOptions?: AlgodClientOptions;
  clientStatic?: typeof DeflyWalletConnect;
  algosdkStatic?: typeof _algosdk;
  network?: Network;
};
