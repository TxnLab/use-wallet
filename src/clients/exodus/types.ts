import { PROVIDER_ID } from "../../constants";
import type _algosdk from "algosdk";
import type { AlgodClientOptions, Network } from "../../types";

export type ClientOptions = {
  onlyIfTrusted: boolean;
};

export type WindowExtended = { exodus: { algorand: Exodus } } & Window &
  typeof globalThis;

export type Bytes = Readonly<Uint8Array>;

export type Exodus = {
  isConnected: boolean;
  address: string | null;
  connect: ({ onlyIfTrusted }: { onlyIfTrusted: boolean }) => Promise<{
    address: string;
  }>;
  disconnect: () => void;
  signAndSendTransaction(transactions: Bytes[]): Promise<{
    txId: string;
  }>;
  signTransaction(transactions: Bytes[]): Promise<Bytes[]>;
};

export type ExodusClientConstructor = {
  client: Exodus;
  id: PROVIDER_ID;
  algosdk: typeof _algosdk;
  algodClient: _algosdk.Algodv2;
  onlyIfTrusted: boolean;
  network: Network;
};

export type InitParams = {
  clientOptions?: ClientOptions;
  algodOptions?: AlgodClientOptions;
  algosdkStatic?: typeof _algosdk;
  network?: Network;
};
