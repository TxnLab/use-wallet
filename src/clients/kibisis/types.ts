import type algosdk from 'algosdk'
import type { Metadata, Network } from '../../types'

export interface Arc0001SignTxns {
  authAddr?: string
  multisig?: string
  signers?: string[]
  stxn?: string
  txn: string
}

export interface Arc0027Account {
  address: string
  name?: string
}

export interface KibisisClientConstructor {
  metadata: Metadata
  algosdk: typeof algosdk
  algodClient: algosdk.Algodv2
  genesisHash: string
  methods: ProviderMethods[]
  network: Network
}

export interface NetworkConfiguration {
  genesisHash: string
  genesisId: string
  methods: ProviderMethods[]
}

export type ProviderMethods =
  | 'enable'
  | 'getProviders'
  | 'postTxns'
  | 'signAndPostTxns'
  | 'signBytes'
  | 'signTxns'

export interface SendRequestWithTimeoutOptions<Params> {
  method: ProviderMethods
  params: Params
  reference: string
  timeout?: number
}

/**
 * message schema
 */

export interface RequestMessage<Params> {
  id: string
  params: Params
  reference: string
}

export interface ResponseMessage<Result = undefined, ErrorData = undefined> {
  error?: ResponseError<ErrorData>
  id: string
  reference: string
  requestId: string
  result?: Result
}

export interface ResponseError<Data = undefined> {
  code: number
  data: Data
  message: string
  providerId: string
}

/**
 * message payloads
 */

export interface EnableParams {
  genesisHash: string
  providerId: string
}

export interface EnableResult {
  accounts: Arc0027Account[]
  genesisHash: string
  genesisId: string
  providerId: string
  sessionId?: string
}

export interface GetProvidersParams {
  providerId: string
}

export interface GetProvidersResult {
  host: string
  icon: string
  name: string
  networks: NetworkConfiguration[]
  providerId: string
}

export interface SignTxnsParams {
  providerId: string
  txns: Arc0001SignTxns[]
}

export interface SignTxnsResult {
  providerId: string
  stxns: (string | null)[]
}
