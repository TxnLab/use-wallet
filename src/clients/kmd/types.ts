import type _algosdk from "algosdk";
import type { AlgodClientOptions, Network } from "../../types";
import { PROVIDER_ID } from "../../constants";

export type ClientOptions = {
  wallet: string;
  password: string;
  host: string;
  token: string;
  port: string;
};

export interface ListWalletResponse {
  id: string;
  name: string;
  driver_name?: string;
  driver_version?: number;
  mnemonic_ux?: boolean;
  supported_txs?: Array<any>;
}

export interface InitWalletHandle {
  wallet_handle_token: string;
  message?: string;
  error?: boolean;
}

export type KMDWalletClientConstructor = {
  client: _algosdk.Kmd;
  id: PROVIDER_ID;
  algosdk: typeof _algosdk;
  algodClient: _algosdk.Algodv2;
  wallet: string;
  password: string;
  network: Network;
};

export type InitParams = {
  clientOptions?: ClientOptions;
  algodOptions?: AlgodClientOptions;
  algosdkStatic?: typeof _algosdk;
  network?: Network;
};
