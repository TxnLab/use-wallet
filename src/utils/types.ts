import { PublicNetwork } from 'src/types'

export const isPublicNetwork = (network: string): network is PublicNetwork => {
  return network === 'betanet' || network === 'testnet' || network === 'mainnet'
}
