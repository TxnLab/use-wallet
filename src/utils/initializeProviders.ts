import type algosdk from "algosdk";
import { PROVIDER_ID, WalletClient, Network } from "../types";
import allClients from "../clients";
import {
  DEFAULT_NODE_BASEURL,
  DEFAULT_NODE_TOKEN,
  DEFAULT_NODE_PORT,
  DEFAULT_NETWORK,
} from "../constants";

type SupportedProviders = { [x: string]: Promise<WalletClient | null> };

type NodeConfig = {
  network: Network;
  nodeServer: string;
  nodeToken?: string;
  nodePort?: string;
};

export const initializeProviders = (
  providers?: PROVIDER_ID[],
  nodeConfig?: NodeConfig,
  algosdkStatic?: typeof algosdk
) => {
  const initializedProviders: SupportedProviders = {};

  const {
    network = DEFAULT_NETWORK,
    nodeServer = DEFAULT_NODE_BASEURL,
    nodePort = DEFAULT_NODE_PORT,
    nodeToken = DEFAULT_NODE_TOKEN,
  } = nodeConfig || {};

  if (!providers || providers.length === 0)
    for (const [id, client] of Object.entries(allClients)) {
      if (id === "kmd") {
        continue;
      }

      initializedProviders[id] = client.init({
        network,
        algodOptions: [nodeToken, nodeServer, nodePort],
        algosdkStatic: algosdkStatic,
      });
    }

  if (providers) {
    for (const id of providers) {
      initializedProviders[id] = allClients[id].init({
        network,
        algodOptions: [nodeToken, nodeServer, nodePort],
        algosdkStatic: algosdkStatic,
      });
    }
  }

  return initializedProviders;
};
