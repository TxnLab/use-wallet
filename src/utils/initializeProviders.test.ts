/**
 * @jest-environment jsdom
 */

/* eslint-disable @typescript-eslint/no-explicit-any */
import algosdk from 'algosdk'
import { initializeProviders } from './initializeProviders'
import allClients from '../clients'
import {
  DEFAULT_NETWORK,
  DEFAULT_NODE_BASEURL,
  DEFAULT_NODE_PORT,
  DEFAULT_NODE_TOKEN,
  PROVIDER_ID
} from '../constants'
import { createMockClient } from '../testUtils/mockClients'
import BaseClient from '../clients/base/base'
import { ClientOptions } from '../types'

describe('initializeProviders', () => {
  beforeEach(() => {
    // Mock the client.init method for all clients
    for (const [id, client] of Object.entries(allClients)) {
      jest.spyOn(client, 'init').mockImplementation(async (params: any) => {
        const providerId = id as PROVIDER_ID
        const clientOptions = params.clientOptions as ClientOptions
        const mockClient = createMockClient(providerId, clientOptions)

        if (mockClient) {
          // Preserve the custom clientOptions from params
          const clientOptions = params.clientOptions
          mockClient.clientOptions = clientOptions

          return Promise.resolve(mockClient)
        }
        return Promise.reject(new Error(`Unsupported provider ID: ${providerId}`))
      })
    }

    // Mock console.warn to avoid polluting the test output
    jest.spyOn(console, 'warn').mockImplementation(() => {
      return
    })
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('should return an empty object when window is undefined (SSR)', async () => {
    const originalWindow = global.window

    Object.defineProperty(global, 'window', {
      get() {
        return undefined
      },
      configurable: true
    })

    const result = await initializeProviders([PROVIDER_ID.EXODUS])
    expect(result).toEqual({})

    Object.defineProperty(global, 'window', {
      value: originalWindow,
      configurable: true
    })
  })

  it('should initialize providers with default node configuration', async () => {
    // Initialize default providers
    const result = await initializeProviders([PROVIDER_ID.EXODUS])

    // Check if the returned object has the correct keys
    expect(Object.keys(result)).toEqual(expect.arrayContaining([PROVIDER_ID.EXODUS]))
    expect(Object.keys(result)).not.toContain(PROVIDER_ID.KMD)
    expect(Object.keys(result)).not.toContain(PROVIDER_ID.MNEMONIC)
    expect(Object.keys(result)).not.toContain(PROVIDER_ID.WALLETCONNECT)
    expect(Object.keys(result)).not.toContain(PROVIDER_ID.ALGOSIGNER)
    expect(Object.keys(result)).not.toContain(PROVIDER_ID.MYALGO)

    // Check if the clients were initialized with the default node configuration
    for (const providerId of Object.keys(result)) {
      expect(allClients[providerId].init).toHaveBeenCalledWith({
        network: DEFAULT_NETWORK,
        algodOptions: [DEFAULT_NODE_TOKEN, DEFAULT_NODE_BASEURL, DEFAULT_NODE_PORT, undefined],
        algosdkStatic: undefined
      })
    }
  })

  it('should return an object with the expected shape', async () => {
    const providers = Object.values(PROVIDER_ID).filter(
      (providerId) => providerId !== PROVIDER_ID.WALLETCONNECT
    )

    // Initialize providers
    const result = await initializeProviders(providers)

    // Check if the returned object has the correct keys
    expect(Object.keys(result)).toEqual(expect.arrayContaining(providers))

    // Check if the returned clients are instances of BaseClient
    for (const clientInstance of Object.values(result)) {
      expect(clientInstance).toBeInstanceOf(BaseClient)
    }
  })

  it('should initialize specified providers only', async () => {
    // Initialize specified providers
    const result = await initializeProviders([PROVIDER_ID.PERA, PROVIDER_ID.DEFLY])

    // Check if the returned object has the correct keys
    expect(Object.keys(result)).toEqual(
      expect.arrayContaining([PROVIDER_ID.PERA, PROVIDER_ID.DEFLY])
    )
  })

  it('should initialize providers using custom node configuration', async () => {
    // Initialize with custom node configuration
    const result = await initializeProviders([], {
      network: 'testnet',
      nodeServer: 'http://localhost',
      nodeToken: 'testToken',
      nodePort: '1234'
    })

    // Check if the clients were initialized with the custom node configuration
    for (const providerId of Object.keys(result)) {
      expect(allClients[providerId].init).toHaveBeenCalledWith({
        network: 'testnet',
        algodOptions: ['testToken', 'http://localhost', '1234'],
        algosdkStatic: undefined
      })
    }
  })

  it('should initialize providers using provided algosdk instance', async () => {
    // Initialize with static algosdk instance
    const result = await initializeProviders([], undefined, algosdk)

    // Check if the clients were initialized with the static algosdk instance
    for (const providerId of Object.keys(result)) {
      expect(allClients[providerId].init).toHaveBeenCalledWith({
        network: DEFAULT_NETWORK,
        algodOptions: [DEFAULT_NODE_TOKEN, DEFAULT_NODE_BASEURL, DEFAULT_NODE_PORT],
        algosdkStatic: algosdk
      })
    }
  })

  it('should initialize providers with a mix of provider IDs and ProviderConfig objects', async () => {
    // Initialize providers with a mix IDs and config objects
    const result = await initializeProviders([
      PROVIDER_ID.DEFLY,
      { id: PROVIDER_ID.PERA, clientOptions: { shouldShowSignTxnToast: false } },
      { id: PROVIDER_ID.MYALGO, clientOptions: { disableLedgerNano: false } },
      PROVIDER_ID.EXODUS
    ])

    // Check if the returned object has the correct keys
    expect(Object.keys(result)).toEqual(
      expect.arrayContaining([PROVIDER_ID.DEFLY, PROVIDER_ID.PERA, PROVIDER_ID.MYALGO])
    )

    // Check if the init method was called with the correct parameters
    expect(allClients[PROVIDER_ID.DEFLY].init).toHaveBeenCalledWith(
      expect.not.objectContaining({
        clientOptions: expect.anything()
      })
    )
    expect(allClients[PROVIDER_ID.PERA].init).toHaveBeenCalledWith(
      expect.objectContaining({
        clientOptions: { shouldShowSignTxnToast: false }
      })
    )
    expect(allClients[PROVIDER_ID.MYALGO].init).toHaveBeenCalledWith(
      expect.objectContaining({
        clientOptions: { disableLedgerNano: false }
      })
    )
    expect(allClients[PROVIDER_ID.EXODUS].init).toHaveBeenCalledWith(
      expect.not.objectContaining({
        clientOptions: expect.anything()
      })
    )
  })
})
