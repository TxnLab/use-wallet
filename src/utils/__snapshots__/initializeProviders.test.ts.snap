// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializeProviders should initialize default wallets with default node configuration 1`] = `
[
  "pera",
  "myalgo",
  "defly",
  "exodus",
  "algosigner",
  "walletconnect",
]
`;

exports[`initializeProviders should initialize specified wallets only 1`] = `
[
  "pera",
  "defly",
]
`;
